src/fu-device.c

```diff
@@ -848,6 +848,24 @@ fu_device_add_counterpart_guid (FuDevice *self, const gchar *guid)
 	fwupd_device_add_guid (FWUPD_DEVICE (self), guid);
 }
 
+/**
+ * fu_device_add_flag:
+ * @self: A #FuDevice
+ * @flag: the #FwupdDeviceFlags
+ *
+ * Adds a specific device flag to the device.
+ *
+ * Since: 1.2.4
+ **/
+void
+fu_device_add_flag (FuDevice *self, FwupdDeviceFlags flag)
+{
+	/* we'll need this until everyone is using a new gnome-software */
+	if (flag & FWUPD_DEVICE_FLAG_NEEDS_SHUTDOWN)
+		flag |= FWUPD_DEVICE_FLAG_NEEDS_REBOOT;
```

Comment:
:(
In this case maybe it's better to do this in the `libfwupd`.  Then next time that ABI bumps and you have a guarantee that gnome-software needs new ABI then you know this can be dropped?

Reply:
I thought the opposite; if we do it in the library then the behaviour becomes effective API...

Reply:
How will you ever know that you have a new gnome-software or a client that supports this then?  I feel like this is effectively a permanent change if you can't tell what client you're talking to.

Reply:
I guess I'm trading API stability of the library with API stability of the daemon. If you think it's better in the lib then I can switch it around, this isn't something I'm passionate about :)

Reply:
Well I believe in when there are comments that something should be temporary there should be a timeline to when it's removed or a criteria to understand when it's to be removed.

If there isn't:
1. Other applications that link can start to depend on a particular behavior
2. There are too many corner cases to worry about that will cause it to never be removed.
3. We're going to forget about it and look at this code 2 years from now, pull up this PR, pull up this discussion and revisit it.

I think in this particular case it's not worth "breaking" ABI yet, I just want to make sure we have a means to undo it at some point.

Reply:
I thought about this some more and given how easy it is to drop a snap into an environment with an older gnome software this is going to become permanent.

I think the right solution is going to be to just drop the comment and let this become permanent API.

Reply:
I also spent several hours thinking about it; I'm proposing dropping the "inheritance" of `NEEDS_REBOOT` for `NEEDS_SHUTDOWN` entirely. Logically, a shutdown does a reboot, but it's a step further. But, it's not really true if you say "this hardware needs a shutdown" and then doing a reboot. I think in this case the drives that need powering down should **only** set `NEEDS_SHUTDOWN` and not `NEEDS_REBOOT`. In the case of an old gnome-software the reboot is actually _worse_ than doing nothing, as the user was told it would be done on next boot rather than not being told anything.

The only way of making sure people actually shutdown with older versions of gnome-software is using a prominent statement in the update description, possibly even the first line so it shows up by default in the updates panel.

Reply:
Okay sounds good.